chain/chain/src/tests/simple_chain.rs

```diff
@@ -24,10 +24,15 @@ fn empty_chain() {
     assert_eq!(chain.head().unwrap().height, 0);
     let hash = chain.head().unwrap().last_block_hash;
     // The hashes here will have to be modified after each change to genesis file.
+    // Same below. You will need to run this test suite at least 4 times to find all the
+    // right hashes. I bet you love it. Now contemplate whether your change actually was
+    // supposed to change these hashes, because that's what change detectors like this one are for.
+    // https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html
+    // Just saying.
```

Comment:
This comment adds a value (more than the test itself), because thanks to it we won't have to have this discussion again. As I have explained, the TotT does apply in this situation, you just don't want to acknowledge it. And it is more than 4, due to the misleading conditional compilation. I've added a comment for that too for people who stumble upon the same problem in the future. Only improvement I see here is to delete the test, or at least the assert.

Reply:
like, seriously, to make this test pass for every flag combination, you'd have to hardcode exponentially many hashes.

Reply:
It doesn’t.  It would add value if you explained how to run the tests to be able to update the hashes.  And also explain that if the hashes changed without commit changing protocol version or any protocol features than this is an issue with the code.

As for running the test four times, you can refactor build_chain such that it checks both hashes at once which reduces number of times the test needs to be run.

Finally, the test is a piece of code which has only one correct result.  If you change the protocol version you change what that correct result is and thus need to update the test.  If you don’t change the protocol version, you can perform any refactoring you want to nearcore† and the test will continue to work.

Such tests are nothing special and are common at Google as well.  A quite visual example are user interface tests which verify that elements on a page have not moved after refactoring CSS, JavaScript or HTML code.  And if you make changes which intend to move the elements you than need to update the tests.

† There is one caveat here in that the use of mock clock does make this a change-test.  But this is not the thing you’re complaining about so I’m ignoring it in this discussion.

Reply:
> like, seriously, to make this test pass for every flag combination, you'd have to hardcode exponentially many hashes.

Your statement is true but is also irrelevant since that’s not what we’re doing.  We’re only testing two combinations which correspond to stable and nightly executables.  That is a bit confusing and I do question sometimes why we have ‘nightly_protocol’ and ‘nightly_protocol_features’ as separate features but it’s irrelevant to this test.

Reply:
> Finally, the test is a piece of code which has only one correct result.

This specific piece of code obviously doesn't, because the result depends on the combination of the feature flags - depending on how you compile it, it will either pass or not.

Reply:
> Finally, the test is a piece of code which has only one correct result. If you change the protocol version you change what that correct result is and thus need to update the test. If you don’t change the protocol version, you can perform any refactoring you want to nearcore† and the test will continue to work.

Ok, so now I think I get it what you are trying to achieve here. The way I know how to handle invariants like that is:
- create a script/binary which outputs a hash and a version
- run this binary on presubmit on master and on the new commit
- compare the results.

Reply:
OK, I see what you mean by the script you’ve described.  It’s certainly something that could be discussed (though I’m currently not convinced) bur regardless of whether we want to change how this is tested, this is outside of the scope of this PR and this PR should just bump the version.

Reply:
removed everything, but the version bump